### PR TITLE
Block serialization: correctly compare default attribute values

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes a little bug in block serialization that I accidentally discovered when working on #53470. When a block declares its attributes, in the `block.json` file, it can also supply a default value. The block serialization, then, when an attribute value equals to the default value, will omit that attribute from the serialized form.

This doesn't work well when the default value is something like empty array (`[]`). Different empty arrays are not equal to each other, `[] !== []`. It's better to compare their serialized forms, i.e., the strings produced by `JSON.stringify`. That's what this PR does.

**How to test:**
There are two unit tests, for the Gallery and Video blocks, that are affected by this.

### CODE REVIEW

File: packages/block-library/src/gallery/block.json [hunk 1]
```diff
@@ -61,10 +61,10 @@
 		},
 		"shortCodeTransforms": {
 			"type": "array",
-			"default": [],
 			"items": {
 				"type": "object"
-			}
+			},
+			"default": []
```
Comment (jsnajdr) [inline packages/block-library/src/gallery/block.json:67]: This is just an aesthetic drive-by change. The intent is to consistently declare the `default` as the last field. See, for example, the `ids` definition that precedes the `shortCodeTransforms` one.

**Review by tyxla — APPROVED** [review]
Great find @jsnajdr, LGTM 🚀

File: packages/blocks/src/api/serializer.js [hunk 2]
```diff
@@ -238,7 +238,8 @@ export function getCommentAttributes( blockType, attributes ) {
 			// Ignore default value.
 			if (
 				'default' in attributeSchema &&
-				attributeSchema.default === value
+				JSON.stringify( attributeSchema.default ) ===
```
Comment (gziolo) [inline packages/blocks/src/api/serializer.js:241]: It would be useful to add test coverage as it might be a bigger issue when dealing with non-empty objects and arrays.

Example:

Screenshot 2023-08-10 at 15 19 26

**Comment by gziolo** [comment]
Great discover. Thank you for working on the fix.

File: packages/blocks/src/api/serializer.js (hunk 2 above)
Comment (jsnajdr) [reply packages/blocks/src/api/serializer.js:241]: The test coverage is there, the `test/integration/full-content` suite does almost 300 tests on various block that are all about parsing and serialization.

Good point about the `{ a, b } === { b, a }` issue. Instead of `JSON.stringify`, we could probably use something like `fast-deep-equal` that sorts object keys before comparing them.

In practice, however, default attribute values tend to be empty-ish objects, and serialization wants to omit values that haven't been modified at all from the default value. So, pragmatically, the `JSON.stringify` solution is likely to be pretty good.

Comment (gziolo) [reply packages/blocks/src/api/serializer.js:241]: > In practice, however, default attribute values tend to be empty-ish objects, and serialization wants to omit values that haven't been modified at all from the default value. So, pragmatically, the `JSON.stringify` solution is likely to be pretty good.

We have an example in the core block (it also applies to its variations, but we override the default value in that case):

https://github.com/WordPress/gutenberg/blob/1bb701445b5ccd45905c0fd1b4d2ffd7395abf4c/packages/block-library/src/query/block.json#L15-L29

However, it's a common issue that plugins encounter, example:

https://github.com/dutchigor/wordpress-block-supports-render-callback-test/blob/3cbef5c6b8cbe95e10979eff5546d10fe5bdb7bb/src/block.json#L21-L28

More details in the following issues:

- https://github.com/WordPress/gutenberg/issues/52600
- https://github.com/WordPress/gutenberg/issues/50229
- https://github.com/WordPress/gutenberg/issues/50027
- https://github.com/WordPress/gutenberg/issues/7342
